lib/inertia_rails/middleware.rb

```diff
@@ -3,41 +3,78 @@ class Middleware
     def initialize(app)
       @app = app
     end
-  
+
     def call(env)
+      @env = env
+
       status, headers, body = @app.call(env)
       request = ActionDispatch::Request.new(env)
 
       ::InertiaRails.reset!
-  
-      return [status, headers, body] unless env['HTTP_X_INERTIA'].present?
-  
-      return force_refresh(request) if stale?(env['REQUEST_METHOD'], env['HTTP_X_INERTIA_VERSION'])
-  
-      if is_redirect_status?(status) &&
-          is_non_get_redirectable_method?(env['REQUEST_METHOD'])
-        status = 303
-      end
-  
-      [status, headers, body]
-    end
-  
+
+      # Inertia errors are added to the session via inertia_redirect_to 
+      request.session.delete(:inertia_errors) unless keep_inertia_errors?(status)
+
+      status = 303 if inertia_non_post_redirect?(status)
+
+      return stale_inertia_get? ? force_refresh(request) : [status, headers, body]
+    end
+
     private
-  
-    def is_redirect_status?(status)
+
+    def keep_inertia_errors?(status)
+      redirect_status?(status) || stale_inertia_request?
+    end
+
+    def stale_inertia_request?
+      inertia_request? && version_stale?
+    end
+
+    def redirect_status?(status)
       [301, 302].include? status
     end
-  
-    def is_non_get_redirectable_method?(request_method)
+
+    def non_get_redirectable_method?
       ['PUT', 'PATCH', 'DELETE'].include? request_method
     end
-  
-    def stale?(request_method, inertia_version)
-      sent_version = InertiaRails.version.is_a?(Numeric) ? inertia_version.to_f : inertia_version
-      saved_version = InertiaRails.version.is_a?(Numeric) ? InertiaRails.version.to_f : InertiaRails.version
-      request_method == 'GET' && sent_version != saved_version
+
+    def inertia_non_post_redirect?(status)
+      inertia_request? && redirect_status?(status) && non_get_redirectable_method?
+    end
+
+    def stale_inertia_get?
+      get? && stale_inertia_request?
     end
-  
+
+    def get?
+      request_method == 'GET'
+    end
+
+    def request_method
```

Comment:
I like these new methods; super clear to read. Since they now depend on the `@env` instance variable being set from within the `call` method, do you think we could put them under a `protected` scope to avoid calling them without that initialization?

Reply:
to be honest, i really don't know... i believe initialization of the middleware happens when the server starts, and the only public method is `call` which is called through the the magic of the framework. i'm not sure how we could run into trouble using `@env` here, but i'm not a middleware guru by any stretch. is there a specific scenario you have in mind? or more that protected would just be following the best practice here?

Reply:
Yeah, definitely not middleware specific, I was just thinking of this object as a plain class. Since we can instantiate the object without an `env`, I thought it would be good to ensure those methods would never be called outside of the `call` method. But, your point is solid that this is a middleware specific class that isn't meant to be instantiated anywhere else.

Reply:
i agree it feels weird to see instance variables being set outside of `initialize`